### PR TITLE
Add new tests to cover some corner cases

### PR DESCRIPTION
**Pull request by ricardotulio**
## Description
This pull request add tests to cover some error scenarios.

## Changes
Add tests to validate when:
- Try to create a user without a username
- Try to create a user with an invalid username
- Try to get a user that doesn't exists

### CODE REVIEW

File: src/main/java/com/ministore/api/controllers/GetUserHandler.java [hunk 1]
```diff
@@ -23,7 +26,13 @@ public GetUserHandler(UserRepository repository) {
...
     )
     public Optional<User> getUser(@PathVariable String id) {
-        return this.repository.findById(id);
+        Optional<User> user = this.repository.findById(id);
+
+        if (user.isEmpty()) { 
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "entity not found");
+        }
+
+        return user;
```
Comment (mverardo) [inline src/main/java/com/ministore/api/controllers/GetUserHandler.java]: There is an interesting case here.
`this.repository.findById()` returns an `Option<User>`, but if you are always checking if the `user.isEmpty` before returning, then `getUser()` doesn't really return an `Option<User>` anymore. It either returns an `User` or throws the exception.

It is important to change the return type of `getUser()` to `User`, because this signals to callers of this function that they don't need to deal with the case where this function returns an empty `Option`.

File: src/main/java/com/ministore/api/controllers/GetUserHandler.java [hunk 2]
```diff
@@ -23,7 +26,13 @@ public GetUserHandler(UserRepository repository) {
         produces = "application/json"
     )
     public Optional<User> getUser(@PathVariable String id) {
-        return this.repository.findById(id);
+        Optional<User> user = this.repository.findById(id);
+
+        if (user.isEmpty()) {
```
Comment (mverardo) [inline src/main/java/com/ministore/api/controllers/GetUserHandler.java]: A good alternative would be to remove this `if` check and use the [`user.orElseThrow()`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElseThrow-java.util.function.Supplier-) method instead.

Let me know if you have any questions on how to do it.

File: src/test/java/com/ministore/api/integration/CreateUserTest.java [hunk 3]
```diff
@@ -51,4 +51,39 @@ void mustBeAbleToCreateAUser() throws Exception
...
+
+    @Test
+    void mustReturnBadRequestWhenPayloadDoesntHaveUsername() throws Exception {
+        this.mockMvc.perform(
+            post("/user")
+            .contentType(MediaType.APPLICATION_JSON)
+            .content("{}")
+            .accept(MediaType.APPLICATION_JSON))
+            .andExpect(status().isBadRequest());
+    }
```
Comment (mverardo) [inline src/test/java/com/ministore/api/integration/CreateUserTest.java:116]: Nice tests!

I'd like to ask you to think about two new cases, and add tests accordingly to cement the behavior:

* What should happen when the same username is sent more than once? i.e. what does the system do if we make the `POST` request with body `"{\"username\":\"ricardotulio\"}"` twice?
* What should happen when the request body includes more fields than the username? i.e.: what does the system do if we receive a body like this: `"{\"username\":\"some.username\", password: \"some password\"}"`?

**Review by mverardo — CHANGES_REQUESTED** [review]
Nice PR! 🔥 

Besides the points I've added on other comments, It might be a good idea for us to talk about how the database is being managed: did you run simple scripts directly on the database? Or are you using some kind of Migration tool to create tables for you?

File: src/test/java/com/ministore/api/integration/CreateUserTest.java [hunk 4]
```diff
@@ -51,4 +51,39 @@ void mustBeAbleToCreateAUser() throws Exception
...
+            post("/user")
+            .contentType(MediaType.APPLICATION_JSON)
+            .accept(MediaType.APPLICATION_JSON))
+            .andExpect(status().isBadRequest());
+    }
+
+    @Test
+    void mustReturnBadRequestWhenUsernameIsInvalid() throws Exception {
+        this.mockMvc.perform(
+            post("/user")
```
Comment (mverardo) [inline src/test/java/com/ministore/api/integration/CreateUserTest.java]: Ah, just another nitpick I should have seen earlier:

Rest URIs represent collections of resources on the server, so they should be written as plural. So the right URI here would be `/users` instead of `/user`.

File: src/main/java/com/ministore/api/controllers/GetUserHandler.java (hunk 1 above)
Comment (ricardotulio) [reply src/main/java/com/ministore/api/controllers/GetUserHandler.java]: Awesome! I applied your suggestion from the previous comment about using `.orElseThrow` and implemented the same approach here.

File: src/main/java/com/ministore/api/controllers/CreateUserHandler.java [hunk 5]
```diff
@@ -32,7 +33,15 @@ private User createUser(@RequestBody CreateUserRequest request)
             throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "username is invalid");
         }
 
-        return this.userService.createUser(request.username);
+        User user = null;
+
+        try {
+            user = this.userService.createUser(request.username);
+        } catch(DataIntegrityViolationException exception) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "username is invalid");
```
Comment (ricardotulio) [inline src/main/java/com/ministore/api/controllers/CreateUserHandler.java]: Some comments:
- Http Status Code 409 maybe is a better option to this scenario

File: src/test/java/com/ministore/api/integration/CreateUserTest.java (hunk 3 above)
Comment (mverardo) [reply src/test/java/com/ministore/api/integration/CreateUserTest.java:116]: Following our conversation today, can you add a test that checks the second case?
